### PR TITLE
Implement MD dispatch to PSA

### LINKED ISSUE #6990: Enable dispatch to PSA in MD (for hashes)
See https://github.com/gilles-peskine-arm/mbedtls/blob/psa-migration-md-cipher-strategy/docs/architecture/psa-migration/md-cipher-dispatch.md#support-for-psa-in-the-md-context and following

Note to self: compared to prototype, need to check if the drivers are initialized (create internal function for that).

### PR DESCRIPTION
**Pull request by mpg**
## Description

Resolves #6990 

See https://github.com/Mbed-TLS/mbedtls/blob/development/docs/architecture/psa-migration/md-cipher-dispatch.md#md-light specifically sub-sections "MD algorithm support macros" to "Error code conversion". (The first subsection was #7120.)

The first few commits just port the implementation from Gilles's [original prototype](https://github.com/Mbed-TLS/mbedtls/pull/6474), with a few related additions (for example, handle size macros at the same time as utility functions). Then changes between the original prototype and the final design document (that is, the dynamic aspect of dispatch) are implemented.

## Gatekeeper checklist

- [x] **changelog** not required - will be added at the end of the series
- [x] **backport** not required - new feature
- [x] **tests** provided

### CODE REVIEW

**Review by mprse — COMMENTED** [review]
Changes and tests looks good. Left 2 comments with for places that needs clarification.

File: tests/include/test/psa_crypto_helpers.h
Comment (mpg) [reply tests/include/test/psa_crypto_helpers.h:321]: I think in builds without `MBEDTLS_PSA_CRYPTO_C`, we're not overwriting anything as those macros have not been defined before. (Evidence: before I added that, I got "undeclared function" warnings, and also if we were re-defining the macros I think we'd get warnings about it.)

But I take your point that the way it's done is not clear enough and I agree that what you're suggesting is much clearer. I'll rework that.

File: tests/suites/test_suite_md.function
Comment (mpg) [reply tests/suites/test_suite_md.function:349]: Good point! I agree that right now those two cases are never executed, as we have only one component where SHA256 & SHA224 are driver only, which is `test_psa_crypto_config_accel_hash_use_psa` which currently doesn't have MD-light enabled, so is skipping this entire test suite.

However, that's temporary and the next PR (#7271) will have MD-light enabled in that component. Also, since things are similar for all hashes, and the test cases for other hashes are executed, I'd say the test gap, while real, is not very worrying. So, would you agree it's acceptable to merge this PR with this small test gap considering it's going to be fixed very soon? Otherwise I can add a component for that (like I did with `accel_hash_keep_builtins`). Wdyt?

Comment (mprse) [reply tests/suites/test_suite_md.function:349]: Yes, I think it is harmless. This can be fixed in https://github.com/Mbed-TLS/mbedtls/pull/7271.

**Review by AndrzejKurek — CHANGES_REQUESTED** [review]
Some comments left.
Also: Can the decision on where to use PSA_INIT() in tests and where not be somehow justified, explained, or highlighted? What I mean is, we have a couple of tests in `test_suite_md` (and in fact in other test suites using md explicitly or under the hood too) that will now dispatch the calls to different places based on `PSA_INIT()`. Should we identify these places to find if there are any that specifically shouldn't use `PSA_INIT()`? Starting from `test_suite_md` itself, tests like md_text, md_hex - why shouldn't these be run with `PSA_INIT()`?

File: library/md.c
Comment (mpg) [reply library/md.c:216]: Thanks for the info, I didn't noticed this had been merged! If that's OK with you, I'd like to leave that for the "size optimization" part - I've added a note about it to an existing issue: https://github.com/Mbed-TLS/mbedtls/issues/6993#issuecomment-1467742073

File: library/psa_crypto.c
Comment (mpg) [reply library/psa_crypto.c:7119]: I think this part should be updated - it probably should have been worded "The _might be_ a problem" because IIRC it's not that we were sure there was a problem, more that there was something to do and we weren't sure it wouldn't cause issues. Now it's done and have not been causing issues.

Comment (mpg) [reply library/psa_crypto.c:7119]: Ah, that's in `strategy.md`, not `md-cipher-dispatch.md`. In that case, if you don't mind I'd rather fix that as part of #7156 because there's potentially a lot more that's out-of-date in that file right now. I've left a note.

Comment (mpg) [reply library/psa_crypto.c:7119]: Currently initializing just the drivers is not supported (`psa_driver_wrapper_init()` is a private function), the only thing that's supported is initializing everything with `psa_crypto_init()` which already has a guard, so I think that's OK.

(Note: https://github.com/Mbed-TLS/mbedtls/pull/6636 will change that but it's not in yet.)

**Comment by mpg** [comment]
@mprse @AndrzejKurek Thanks for your reviews! I think I've addressed all of your feedback, either by changes in the code, or making a note to address it later if that's OK with you, or in one instance explaining why I think no change's needed.

> Can the decision on where to use PSA_INIT() in tests and where not be somehow justified, explained, or highlighted?

Ah, I think I just added PSA_INIT() when a test would otherwise fail, which is a very weak rationale now that you make me spell it out :) Indeed when thinking about it, all tests that call `md_setup()` or do an actual computation now need `PSA_INIT()` at least in some configurations. I've now added that in a systematic way, thanks for pointing it out.

File: library/psa_crypto.c
Comment (AndrzejKurek) [reply library/psa_crypto.c:7119]: What's the gain of introducing `drivers_initialized` then? I assumed that it should support partial `psa_crypto_init` success:
- If `psa_crypto_init` succeeds, then indeed, `global_data.initialized` is equal to 1 and there's no need to check `drivers_initialized`;
- If `psa_crypto_init` fails after `psa_driver_wrapper_init` was called succesfully then the state is inconsistent and we could prevent a second call to drivers that were already initialized, during a second call to `psa_crypto_init`;
If that's not the point and we don't want to support partial initialization, then why add `drivers_initialized` at all?

Comment (mpg) [reply library/psa_crypto.c:7119]: Ah, excellent question! Indeed in the context of this PR there's not point yet, I'm just anticipating on #7155 where the entropy module will start using MD, with the goal of having it work when hashes are provided by drivers. The entropy module is first used at the "init RNG" stage in `psa_crypto_init()`, so when it will call MD, `global_data.initialized` will still be 0 (it's only changed to 1 at the end of the function), but we'll want MD to dispatch to drivers nonetheless (which is also the reason for the change of ordering).

Now that you mention it, my PR structure is not clean: in this PR indeed I shouldn't have changed the order or introduced `drivers_initialized`, but instead use `global_data.initialized` in `psa_can_do_hash()`, and left the rest for the PR about the entropy module. I'm afraid I got a bit ahead of myself here.

Comment (AndrzejKurek) [reply library/psa_crypto.c:7119]: Ah, now it makes sense! Thanks for explaining. I still think that we could optimize to not call init on drivers if they are already initialized, but it's a minor.

File: tests/suites/test_suite_md.function
Comment (AndrzejKurek) [inline tests/suites/test_suite_md.function:8]:
```suggestion
#define MD_PSA_INIT() ((void) 0)
```
Without this change the code will not compile out of the box.

File: include/mbedtls/md.h
Comment (AndrzejKurek) [inline include/mbedtls/md.h:40]: Minor: this applies to all lines below, or maybe it would be just better to say that "PSA can perform xxx", just as with MD above.

**Comment by AndrzejKurek** [comment]
> I've now added that in a systematic way, thanks for pointing it out.
What about other test suites that call it indirectly? ECJPAKE, pkcs12, x509write, ct, RSA (in rsaes_oaep, which aparently isn't tested in `test_suite_rsa`), and ssl all use md. Do we want to also make sure that PSA_INIT is called in these tests? Or do we assume that these should be implementation-agnostic and work in any available way?

**Comment by mpg** [comment]
> > I've now added that in a systematic way, thanks for pointing it out.
> 
> What about other test suites that call it indirectly? ECJPAKE, pkcs12, x509write, ct, RSA (in rsaes_oaep, which aparently isn't tested in `test_suite_rsa`), and ssl all use md. Do we want to also make sure that PSA_INIT is called in these tests? Or do we assume that these should be implementation-agnostic and work in any available way?

I was thinking about that. I think for now they don't need `PSA_INIT` because they're only supposed to work when either hashes are available in software or (1) for PK, X.509, TLS: `USE_PSA_CRYPTO` is defined, (2) for the others: `MD_C` is not defined. Of course the ultimate goal is that these will start working in more circumstances, at which point the calls to PSA_INIT will start being necessary. I'd say that will happen in #6991 when we replace the existing constraining macros from `legacy_or_psa.h` with `MBEDTLS_MD_CAN` and stop doing the `#if defined(MD_C) /* use MD */ #else /* use PSA */ #endif` dance in mixed-domain modules such as RSA.

Note: RSA-OAEP is tested in `test_suite_pkcs1_v21` I think. FWIW I agree it's confusing, and this should be renamed to `test_suite_rsa_pkcs1_v21` at least, to show it's related to RSA.

**Comment by mpg** [comment]
I've labelled "do not merge" as I'd rather wait until after the code freeze for the release to merge this. This is not really complete without #6991, and while I think breaking the work in multiple issues makes sense to avoid overly large PRs, I think in a release it's better to either not have this PR, or have #6991 to go with it.

**Review by AndrzejKurek — DISMISSED** [review]
Looks good given the constraints discussed in comments and planned follow-up work.

**Comment by gilles-peskine-arm** [comment]
> Can the decision on where to use PSA_INIT() in tests and where not be somehow justified

`PSA_INIT` just calls `psa_crypto_init`. It needs to be called in any test case that uses PSA but does not call `psa_crypto_init` explicitly, and must not be called in test cases that must run when `MBEDTLS_PSA_CRYPTO_C` is disabled. I don't remember any compelling reason to use `PSA_INIT` rather than `psa_crypto_init`, it's mainly there for symmetry with `PSA_DONE` and also in case it needed to do more in the future.

`PSA_DONE` calls `mbedtls_psa_crypto_free`, but before that it checks that the key store is empty (otherwise the test is marked as failed: this is a resource leak), and purges the on-disk storage (otherwise a test case that leaks persistent keys could perturb subsequent test cases). It needs to be called in any test case that uses PSA and does not have a good reason to bypass the normal resource leak check (typically, test cases around PSA session management, e.g. testing that persistent keys are actually persistent).

**Comment by mpg** [comment]
> Functions in e.g. RSA never knowingly call PSA code. They call the MD module, and maybe it will call PSA, but that's a problem for MD to solve, it can't assume that its caller has set up anything. So RSA tests should not call `PSA_INIT`. They need to work whether PSA crypto is initialized or not.
> 
I almost agree but not quite: in builds where SHA-xxx is provided only by a driver, RSA with SHA-xxx can only work if `psa_crypto_init()` has been called. And that's not a compatibility problem because:
- in 3.3, RSA can only use drivers when `MD_C` is compiled out, and the doc says that in those builds you must call `psa_crypto_init()`;
- prior to 3.3, RSA with SHA-xxx would fail when SHA-xxx was driver-only, regardless of whether `psa_crypto_init()` had been called.

I agree that most of the time we should test things like RSA without calling `psa_crypto_init()`, but in some builds we'll need to call it. What I intend to do is call it only when `MBEDTLS_MD_SOME_PSA` is defined. That's not the most accurate, since is doesn't differentiate by hash, but:
- `test_suite_md` has accurate tests, one for each hash;
- making all the RSA tests (and other similar modules) decide whether to call `psa_crypto_init()` based on the hash being used would be a lot of work with IMO a low usefulness (considering the previous point), so I'd rather keep things simple.

See #7284 (preview) and in particular [this commit](https://github.com/Mbed-TLS/mbedtls/pull/7284/commits/f278b1648834649499959796a15d6babb1de2fe9) and [the next one](https://github.com/Mbed-TLS/mbedtls/pull/7284/commits/bb6c6151d93159b0d5a7793afc5f5b3738e9bd0d) - using ECJPAKE as an example because it's a pretty simple one.

> And we should make sure we have some test cases that run in builds with `MBEDTLS_PSA_CRYPTO_C`, that call md-light functions (directly or indirectly) without having called `psa_crypto_init`, and that would fail if md-light was incorrectly calling `psa_hash_xxx`.
> 
We have such tests in `test_suite_md`, but for other modules, "that would fail if md-light was incorrectly calling `psa_hash_xxx`" seems hard to achieve: currently `psa_hash_xxx()` succeeds even if `psa_crypto_init()` hasn't been called, so if we wanted to really ensure that, we'd need to develop some instrumentation for that purpose.

> PK, X509 and SSL tests should call `USE_PSA_INIT` if they perform any cryptography.

Yes, that's the simple case.

**Comment by gilles-peskine-arm** [comment]
> I almost agree but not quite: in builds where SHA-xxx is provided only by a driver, RSA with SHA-xxx can only work if psa_crypto_init() has been called.

Ah, right, I'd completely missed that case yesterday.

> test_suite_md has accurate tests, one for each hash;

Meaning there is a test case calling `mbedtls_md(MBEDTLS_MD_xxx)` without `psa_crypto_init` that runs whenever `MBEDTLS_MD_xxx` is enabled via `{md,sha}*.c`? Ok, I agree that's sufficient, we don't need the full complexity in higher-level modules.

>  "that would fail if md-light was incorrectly calling psa_hash_xxx" seems hard to achieve: currently psa_hash_xxx() succeeds even if psa_crypto_init() hasn't been called, so if we wanted to really ensure that, we'd need to develop some instrumentation for that purpose.

That's a deficiency in the test drivers — we aren't testing driver init. Eventually we should reach a point where the test drivers fail if their init hasn't been called.

Thanks for writing all this out — I found it very useful.

**Comment by gilles-peskine-arm** [comment]
Note, there's a merge conflict.

**Comment by mpg** [comment]
> Meaning there is a test case calling `mbedtls_md(MBEDTLS_MD_xxx)` without `psa_crypto_init` that runs whenever `MBEDTLS_MD_xxx` is enabled via `{md,sha}*.c`? Ok, I agree that's sufficient, we don't need the full complexity in higher-level modules.
> 
Exactly, that's what we have in this PR.

> That's a deficiency in the test drivers — we aren't testing driver init. Eventually we should reach a point where the test drivers fail if their init hasn't been called.
> 
Agreed, that's something we should add at some point.

> Thanks for writing all this out — I found it very useful.

Well, thanks for your review!

**Comment by mpg** [comment]
Rebased and force-pushed to resolve the conflict (a trivial one: development had whitespace changes in the definition of `PSA_DONE()` which I'm moving to another place in the same file here).

**Review by valeriosetti — APPROVED** [review]
LGTM!
I only found one possible suggestion about the mbedtls_md_error_from_psa() function, but then I saw it was [already discussed](https://github.com/Mbed-TLS/mbedtls/pull/7242#discussion_r1133910696) and found a plan for it, so I'm OK with this PR.
Just wait for the CI to be OK as well.

File: tests/suites/test_suite_md.function
Comment (AndrzejKurek) [inline tests/suites/test_suite_md.function:131]: Very minor:
```suggestion
    /* Note: PSA Crypto init not needed for info functions */
```

**Review by AndrzejKurek — APPROVED** [review]

File: tests/suites/test_suite_md.function
Comment (mpg) [reply tests/suites/test_suite_md.function:131]: I'll fix that in a follow-up in order to avoid another round of CI here.

Comment (mpg) [reply tests/suites/test_suite_md.function:131]: https://github.com/Mbed-TLS/mbedtls/pull/7271/commits/6ea8d3414fc835b5490ef048a419ac2c5454dc2a

**Pull request merged**
